### DISCUSSION TITLE
How can I convert geodetic (latitude,longitude,altitude) coordinates into utm (x,y,z) coordinate

### DISCUSSION THREAD

**Discussion by xleEmesent** [pyproj4/pyproj#958] [Q&A]
All:

How can I convert geodetic (latitude,longitude,altitude) coordinates into utm (x,y,z) coordinate with pyproj or any python tool? Currently I have tried 
```
latAngle=-16.7249452
longAngle= 128.37154429999998
altAngle= 0.022959

# expected to have
# x = 433007.2008822106
# y = 8150766.755038355
# z is around 230.8296793483522

p = Proj(proj='utm',zone=52,ellps='WGS84', preserve_units=False,south=True)
x,y = p(128.37154429999998, -16.7249452)
>>> x
433007.2008822106
>>> y
8150766.755038355
```
It works only with two parameters, (longAngle,latAngle). Is there anyway to plug in the 3rd parameter altAngle into the function as well?

as:
```
>>> x,y,z = p(128.37154429999998, -16.7249452,0.022959)
Traceback (most recent call last):
ValueError: not enough values to unpack (expected 3, got 2)
```
I also tried:
```
p.transform(128.37154429999998, -16.7249452, 0.022959)
>>> p.transform(128.37154429999998, -16.7249452, 0.022959)
(433007.2008822106, 8150766.755038355, 0.022959)
``` 
The z corresponds altAngle is not transformed as expected, still 0.022959

Many Thanks!

**Comment by snowman2** [comment]
I recommend using the `Transformer`. Also see: 
- https://pyproj4.github.io/pyproj/stable/gotchas.html#proj-not-a-generic-latitude-longitude-to-projection-converter
- https://gis.stackexchange.com/questions/409379/convert-gps-coordinates-into-3-dimensional-cartesian-coordinates-x-y-z
- https://gis.stackexchange.com/questions/391224/postgis-st-transform-seems-to-treat-the-vertical-datum-but-not-the-equivalent-py/392232#392232

**Reply by xleEmesent** [reply]
@snowman2 Sorry, I am fresh to this area. If I intend to use Transformer class as following.
What are the parameters for CRS objects I should use here? Can you shed more light on this? 
(As the destination is in utm 52k)

![image](https://user-images.githubusercontent.com/66909631/134091856-cbf0dabb-a6d0-4975-8602-b7f10a7ab0b7.png)

```
from pyproj import CRS
crsFrom = CRS("???") # what's the parameter?
crsTo = CRS("???"") # what's the parameter?
from pyproj import Transformer
transformer = Transformer.from_crs(crsFrom, crsTo)
latAngle=-16.7249452
longAngle= 128.37154429999998
altAngle= 0.022959
transformer.transform(latAngle, longAngle,altAngle)
```

**Reply by snowman2** [reply]
> The altAngle[0] is not transformed as expected

What is your expected result?

**Reply by snowman2** [reply]
Added more references in the answer that may be useful.

**Reply by xleEmesent** [reply]
Sorry, @snowman2 ,Alan. It should The z values corresponds to altAngle. Apologise for the confusion.  

```
latAngle=-16.7249452
longAngle= 128.37154429999998
altAngle= 0.022959
```
I am expecting pass those 3 parameters the transform function, results would be 
```
x,y,z = transformer.transform(latAngle, longAngle,altAngle)
# expected to have
# x = 433007.2008822106
# y = 8150766.755038355
# z is around 230.8296793483522
```

**Reply by snowman2** [reply]
If your Z component of your input is an angle, you may have success with:
https://pyproj4.github.io/pyproj/stable/build_crs.html

However, I suspect your source projection is geostationary: https://gis.stackexchange.com/questions/345650/how-do-i-add-projection-to-this-netcdf-file-satellite/

**Reply by xleEmesent** [reply]
@snowman2 , Alan. Myy mistake. The input for z is not in angle. It is in meters. The story is that I have two sensors mounted on the same devices. The 1st one is from the GPS sensor. It gives me the measurement in geodetic (latitude,longitude,altitude) coordinates
```
latGPS=-16.7249452
longGPS= 128.37154429999998
altGPS= 0.022959
```
Then 2nd sensor has the position values in utm 52k as 
```
latLidar = 433008.446746611	
longLidar = 8150763.98543988	
altLidar = 230.829679348352
```
I am trying to convert the 1st measurement from GPS senor to the same utm 52k so I can measure the difference between 2 measurement.

It seems that conversion using pyProj for latGPS to latLidar, and longGPS to longLidar is okay. But not on altitude. Thanks.

**Reply by snowman2** [reply]
What is the vertical datum for your lidar data?

**Reply by xleEmesent** [reply]
@snowman2 , had an investigation on the altitude and it turned out that the there is scale bug in GPS device. By scale up all altitude data from GPS device, then it works fine.  Many thanks for your help!!!
